core/src/signal.rs

```diff
@@ -1,20 +1,54 @@
 use crate::action::ActionWrapper;
-use std::{
-    sync::mpsc::{channel, sync_channel, Receiver, SyncSender},
-    thread,
-};
+use crossbeam_channel::{unbounded, Receiver, Sender};
+use holochain_core_types::{error::HolochainError, json::JsonString};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::thread;
 
-#[derive(Debug)]
+#[derive(Clone, Debug, DefaultJson)]
 pub enum Signal {
     Internal(ActionWrapper),
-    User,
+    User(JsonString),
+    // NB: this is part of a temporary hack that will be removed
+    // as soon as a browser light client is implemented!
+    Holo(JsonString),
 }
 
-pub type SignalSender = SyncSender<Signal>;
+impl Serialize for Signal {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Signal::Internal(_) => serializer.serialize_newtype_variant(
+                "Signal",
+                0,
+                "Internal",
+                "(Internal signal serialization not yet implemented)",
+            ),
+            Signal::User(msg) => {
+                serializer.serialize_newtype_variant("Signal", 1, "User", &msg.to_string())
+            }
+            Signal::Holo(msg) => {
+                serializer.serialize_newtype_variant("Signal", 2, "Holo", &msg.to_string())
```

Comment:
What is this Holo Signal?

Reply:
Good catch, this can be removed. It was from back when I was assuming that the wormhole hack would be implemented via Signals